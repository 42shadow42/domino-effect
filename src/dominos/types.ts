import { List, Record, Collection, Map, Set, OrderedMap, OrderedSet, Stack, Seq } from 'immutable'
import { ObservableCache } from 'src/observables/cache'
import { ObservableValueSubscriber } from '../observables'
import { Store } from '../store'

export type Context =
	| List<any>
	| Record<object>
	| Collection<any, any>
	| Map<any, any>
	| Set<any>
	| OrderedMap<any, any>
	| OrderedSet<any>
	| Stack<any>
	| Seq<any, any>
	| string
	| boolean
	| number
	| undefined
export type DominoType = 'standard' | 'trigger'
export type DominoMetadata = { type: DominoType }

export type GetDominoValue<TValue, TContext extends Context> = (context?: TContext) => TValue
export type SetDominoValue<TValue> = (value: TValue) => void
export type DeleteDominoValue = () => void
export type SubscribeDominoValue<TValue> = (
	subscriber: ObservableValueSubscriber<TValue>,
) => void
export type UnsubscribeDominoValue<TValue> = (
	subscriber: ObservableValueSubscriber<TValue>,
) => void

export type DominoUtils<TValue, TContext extends Context> = {
	get: GetDominoValue<TValue, TContext>
	subscribe: SubscribeDominoValue<TValue>
	unsubscribe: UnsubscribeDominoValue<TValue>
	delete: DeleteDominoValue
	debugLabel?: string
}

export type TriggerDominoUtils<TValue, TContext extends Context> = DominoUtils<
	TValue,
	TContext
> & {
	set: SetDominoValue<TValue>
}

export type CoreDomino<TValue, TContext extends Context> = ((
	store: Store,
	context?: TContext,
) => DominoUtils<TValue, TContext>) &
	DominoMetadata
export type TriggerDomino<TValue, TContext extends Context> = ((
	store: Store,
	context?: TContext,
) => TriggerDominoUtils<TValue, TContext>) &
	DominoMetadata

export const isTriggerDomino = <TValue, TContext extends Context>(
	domino: CoreDomino<TValue, TContext> | TriggerDomino<TValue, TContext>,
): domino is TriggerDomino<TValue, TContext> => domino.type === 'trigger'

export type CoreDominoSettings = {
	debugLabel?: string
	ttl?: number
}

export type DominoEffectUtils<TContext extends Context> = {
	get: <TValue, TContext extends Context>(source: CoreDomino<TValue, TContext>, context?: TContext) => TValue
	manage: <TValue, TContext extends Context>(
		trigger: TriggerDomino<TValue, TContext>,
		context?: TContext,
	) => {
		value: TValue
		set: SetDominoValue<TValue>
	},
	context: TContext | undefined
	cache: ObservableCache<any, any>
}

export type DominoDestroyedUtils = {
	cache: ObservableCache<any, any>
}

export type DominoEffectCalculation<TValue, TContext extends Context> = (
	utils: DominoEffectUtils<TContext>,
	context?: TContext,
) => TValue

export type DominoEffectSettings = {
	debugLabel?: string
	ttl?: number
	onDelete?: ({ cache }: DominoDestroyedUtils) => void 
}

export type CacheKey<TContext> = {store: Store, context: TContext | undefined }
