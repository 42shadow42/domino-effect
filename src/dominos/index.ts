export * from './domino'
export * from './trigger'
export * from './types'
export * from './store'
