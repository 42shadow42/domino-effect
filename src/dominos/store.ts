import { Record } from 'immutable';
import { Context } from './types';
import { ObservableMap, ObservableValue } from '../observables'

export type StoreKey = { handle: symbol; context: Context | undefined }
export class Store extends ObservableMap<Record<StoreKey>, ObservableValue<any>> {}

export const STORES = new ObservableMap<symbol, Store>()
export const GLOBAL_STORE = new Store()

STORES.set(Symbol('Global'), GLOBAL_STORE)

export const createStore = (name: string): [Store, symbol] => {
	const store = new Store()
	const symbol = Symbol(name)
	STORES.set(symbol, store)
	return [store, symbol]
}

export const deleteStore = (handle: symbol) => {
	return STORES.delete(handle)
}
