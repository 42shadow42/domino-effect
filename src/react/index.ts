export { useDomino } from './useDomino'
export { useAsyncDomino } from './useAsyncDomino'
export { DominoSuspense } from './suspense'
export type { DominoSuspenseProps } from './suspense'  