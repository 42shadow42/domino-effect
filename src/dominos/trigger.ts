import { Map, List } from 'immutable'
import { ObservableValue, ObservableValueSubscriber } from '../observables'
import { Store } from '../store'
import {
	CoreDominoSettings,
	DominoMetadata,
	TriggerDomino,
	TriggerDominoUtils,
} from './types'

export const trigger = <TValue, TContext = undefined>(
	factory: (context?: TContext) => TValue,
	settings: CoreDominoSettings = {},
): TriggerDomino<TValue, TContext> => {
	const handle = Symbol()

	const { debugLabel, ttl } = settings

	let cache = Map<
		List<Store | TContext | undefined>, TriggerDominoUtils<TValue, TContext>>()
	const metadata: DominoMetadata = { type: 'trigger' }

	return Object.assign((store: Store, context?: TContext) => {
        const cacheKey = List([store, context])
		if (cache.has(cacheKey)) {
			return cache.get(cacheKey)!
		}
        const storeKey = List([handle, context])

        let gcTimeout: NodeJS.Timeout
        const utils: TriggerDominoUtils<TValue, TContext> = Object.assign({
            get: () => {
                if (!store.has(storeKey)) {
                    store.set(storeKey, new ObservableValue(factory(context)))
                }
                // private handle ensures the type will always be of TValue
                return store.get(storeKey)!.get()
            },
            set: (value: TValue) => {
                if (!store.has(storeKey)) {
                    store.set(storeKey, new ObservableValue(value))
                }
                // private handle ensures the type will always be of TValue
                store.get(storeKey)!.set(value)
            },
            subscribe: (subscriber: ObservableValueSubscriber<TValue>) => {
                if (!store.has(storeKey)) {
                    store.set(storeKey, new ObservableValue(factory(context)))
                }
                store.get(storeKey)!.subscribe(subscriber)
                clearTimeout(gcTimeout)
            },
            unsubscribe: (subscriber: ObservableValueSubscriber<TValue>) => {
                if (!store.has(storeKey)) {
                    return
                }
                const count = store.get(storeKey)!.unsubscribe(subscriber)
                if (count === 0 && ttl !== undefined) {
                    gcTimeout = setTimeout(() => {
                        utils.delete()
                    }, ttl)
                }
            },
            delete: () => {
                const newCache = cache.delete(cacheKey)
                const deleted = newCache.size !== cache.size
                cache = newCache
                return store.delete(storeKey) && deleted
            },
            debugLabel,
        }, metadata)

        cache = cache.set(cacheKey, utils)

		return cache.get(cacheKey)!
	}, metadata)
}
