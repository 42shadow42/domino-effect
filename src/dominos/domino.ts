import { ObservableValue, ObservableValueSubscriber } from "../observables"
import { Store } from "../store"
import { CoreDomino, DominoMetadata, DominoUtils, TriggerDomino } from "./trigger"

export type DominoEffectUtils = {
    get: <TValue>(trigger: TriggerDomino<TValue>) => TValue
}

export type DominoEffectCalculation<TValue> = (utils: DominoEffectUtils) => TValue

export type DominoEffectSettings = {
    debugLabel?: string
}

export const domino = <TValue>(calculation: DominoEffectCalculation<TValue>, settings: DominoEffectSettings = {}): CoreDomino<TValue> => {
    const handle = Symbol()

    const { debugLabel } = settings
    const metadata: DominoMetadata = { type: 'standard' }

    const cache = new Map<Store, DominoUtils<TValue>>()

    return Object.assign((store: Store) => {
        if (cache.has(store)) {
            return cache.get(store)!
        }

        let _dependencies = new Set<TriggerDomino<any>>()

        const subscription = () => {
            _dependencies.forEach((dependency) => dependency(store).unsubscribe(subscription))
            _dependencies = new Set<TriggerDomino<any>>()
            // if (!store.has(handle)) {
            //     store.set(handle, new ObservableValue(calculation(utils)))
            // }
            // // private handle ensures the type will always be of TValue
            // return store.get(handle)!.get()
            store.get(handle)?.set(calculation(utils))
        }

        const utils = {
            get: <TValue>(trigger: TriggerDomino<TValue>) => {
                const domino = trigger(store)
                _dependencies.add(trigger)
                domino.subscribe(subscription)
                return domino.get()
            }
        }

        cache.set(store, {
            subscribe: (subscriber: ObservableValueSubscriber<TValue>) => {
                if (!store.has(handle)) {
                    store.set(handle, new ObservableValue(calculation(utils)))
                }
                store.get(handle)!.subscribe(subscriber)
            },
            unsubscribe: (subscriber: ObservableValueSubscriber<TValue>) => {
                if (!store.has(handle)) {
                    return
                }
                store.get(handle)!.unsubscribe(subscriber)
            }, 
            get: () => {
                if (!store.has(handle)) {
                    store.set(handle, new ObservableValue(calculation(utils)))
                }
                // private handle ensures the type will always be of TValue
                return store.get(handle)!.get()
            },
            delete: () => {
                _dependencies.forEach((dependency) => dependency(store).unsubscribe(subscription))
                _dependencies = new Set<TriggerDomino<any>>()
                cache.delete(store)
                return store.delete(handle)
            },
            debugLabel,
        })
        
        return cache.get(store)!
    }, metadata)
}