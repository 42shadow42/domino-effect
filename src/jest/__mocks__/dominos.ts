import { mockDomino } from '../domino'
import { mockTrigger } from '../trigger'
export const core = mockTrigger<string, string>()
export const derivative = mockDomino<string, string>()
