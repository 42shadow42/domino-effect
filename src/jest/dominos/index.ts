import { domino, trigger } from '../../dominos'
export const core = trigger<string, string>(() => 'N/A')
export const derivative = domino<string, string>(() => 'N/A')