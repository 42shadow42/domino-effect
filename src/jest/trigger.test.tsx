/**
 * @jest-environment jsdom
 */

import { Fragment, useEffect } from 'react'
import { useDomino, useManagedDomino } from '../react'
import { Store } from '../dominos'
import { core } from './__utils__/dominos'
import { render, cleanup } from '@testing-library/react'
import { core as mock } from './__mocks__/dominos'

jest.useFakeTimers()

jest.mock('./__utils__/dominos', () => require('./__mocks__/dominos'))

describe('mockDomino', () => {
	const Component = () => {
		const [value] = useDomino(core, { context: 'context' })
		return <Fragment>{value}</Fragment>
	}

	afterEach(() => {
		jest.clearAllMocks()
	})

	it('should mock get', () => {
		mock.get.mockReturnValue('testing')
		const { container } = render(<Component />)

		expect(container).toMatchSnapshot()
		expect(mock.get).toBeCalledWith(expect.any(Store), 'context')
	})

	it('should mock set', () => {
		const Component = () => {
			const utils = useManagedDomino(core, { context: 'context' })

			useEffect(() => {
				utils.set('testing')
			}, [])

			return <Fragment>{utils.get()}</Fragment>
		}

		render(<Component />)
		cleanup()

		expect(mock.set).toBeCalledWith(expect.any(Store), 'context', 'testing')
	})

	it('should mock subscribe', () => {
		render(<Component />)

		expect(mock.subscribe).toBeCalledWith(
			expect.any(Store),
			'context',
			expect.any(Function),
		)
	})

	it('should mock unsubscribe', () => {
		render(<Component />)
		cleanup()

		expect(mock.unsubscribe).toBeCalledWith(
			expect.any(Store),
			'context',
			expect.any(Function),
		)
	})

	it('should mock delete', () => {
		const Component = () => {
			const utils = useManagedDomino(core, { context: 'context' })

			useEffect(() => {
				utils.delete()
			}, [])

			return <Fragment>{utils.get()}</Fragment>
		}

		render(<Component />)
		cleanup()

		expect(mock.delete).toBeCalledWith(expect.any(Store), 'context')
	})
})
