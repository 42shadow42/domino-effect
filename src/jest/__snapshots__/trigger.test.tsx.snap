// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mockDomino should mock get 1`] = `
<div>
  testing
</div>
`;
